### ISSUE TITLE
mdast Inline code should strip prefix+suffix space

### ISSUE THREAD

**Issue by yshavit** [wooorm/markdown-rs#122]
Per the CommonMark spec (and GH-flavored Markdown, and probably others):

> [6.1 Code spans](https://spec.commonmark.org/0.31.2/#code-spans)
> If the resulting string both begins and ends with a [space](https://spec.commonmark.org/0.31.2/#space) character, but does not consist entirely of [space](https://spec.commonmark.org/0.31.2/#space) characters, a single [space](https://spec.commonmark.org/0.31.2/#space) character is removed from the front and back

So, this markdown:

```markdown
`` hello ` ``
```

should result in `` hello ` `` (note: no space before the "hello" or after the trailing backtick).

Instead, markdown-rs (as of 1.0.0-alpha.16) renders it as ``  hello `  ``, with the spaces remaining on either side.

**Comment by yshavit** [comment]
I would be happy to take a crack at fixing this, if you'd like!

**Comment by wooorm** [comment]
Pretty sure it does that. I think you are using the old package, are you using the alpha as mentioned in the readme?

**Comment by wooorm** [comment]
Can you make a reproduction otherwise?

**Comment by yshavit** [comment]
Hm, I'm self-taught and a newbie to Rust, so it could very well be I'm doing something wrong. But my `Cargo.toml` has:

```toml
[dependencies]
markdown = "1.0.0-alpha.19"
```

The following test fails for me, in my project:

```rust
#[test]
fn repro() {
    let markdown = "`` hello ` ``";
    let parsed = markdown::to_mdast(markdown, &ParseOptions::default()).unwrap();
    let Node::Root(root) = parsed else {
        panic!("not a root");
    };
    let Some(Node::Paragraph(paragraph)) = root.children.first() else {
        panic!("no paragraph");
    };
    let Some(Node::InlineCode(inline)) = paragraph.children.first() else {
        panic!("no paragraph");
    };
    assert_eq!(inline.value, "hello `");
}
```

Which file would be an appropriate place for me to try to repro this as a test in the markdown-rs repo?

**Comment by yshavit** [comment]
Oh, interesting -- I do see a test going all the way to html, which has the correct output: [`code_text.rs#L26-L30`][1]:

```rust
assert_eq!(
    to_html("`` foo ` bar ``"),
    "<p><code>foo ` bar</code></p>",
    "should support code w/ more accents"
);
```

Is it possible this is being fixed at the html level, rather than at the Node level?

[1]: https://github.com/wooorm/markdown-rs/blob/main/tests/code_text.rs#L26-L30

**Comment by yshavit** [comment]
(sorry for the comment spam!)

Yes, it looks like that's what's happening. Specifically, the spaces are being trimmed in [`to_html::on_exit_raw_text`][1], but there isn't equivalent code in [`to_mdast::on_exit_raw_text`][2]. Is that where a fix would go?

[1]: https://github.com/wooorm/markdown-rs/blob/1.0.0-alpha.19/src/to_html.rs#L894
[2]: https://github.com/wooorm/markdown-rs/blob/1.0.0-alpha.19/src/to_mdast.rs#L1056

**Comment by wooorm** [comment]
No worries! Thanks for the PR!

So, I can’t find definitive proof in the commits/git blame, but here’s what my guess is:
- in the javascript code, the actual parser does this work
- that isn’t great, so in the rust code, I moved it into the `to_html` code, which is heavily tested, I used this to make sure that it matches CM
- `to_mdast` is less tested, because there aren’t as many mdast tools in rust yet, so I must’ve taken the javascript code (which assumes this is done in the parser) and ported it to rust

I do indeed think `to_mdast::on_exit_raw_text` is where the fix can go.
Similar to this, there’s also the table handling in both places.
There’s a note above it, saying that it would be good to share the code.
But we aren’t there yet!
